### ISSUE TITLE
Zero as null pointer constant

### ISSUE THREAD

**Issue by stertingen** [chriskohlhoff/asio#567]
I build by project with "-Wzero-as-null-pointer-constant" and get annoyed by the following warnings:

```
/home/hermann/turag/programmierung/ros/src/turag_tina/tina/platform/ros/serialfieldbus.cpp: In function ‘typename boost::asio::async_result<typename std::decay<WriteHandler>::type, void(boost::system::error_code, long unsigned int)>::return_type boost::asio::async_read(AsyncReadStream&, const MutableBufferSequence&, ReadHandler&&, typename std::enable_if<boost::asio::is_mutable_buffer_sequence<MutableBufferSequence>::value>::type*) [with AsyncReadStream = boost::asio::basic_serial_port<>; MutableBufferSequence = boost::asio::mutable_buffers_1; ReadHandler = boost::_bi::bind_t<void, boost::_mfi::mf2<void, TURAG::Feldbus::SerialFieldbus, const boost::system::error_code&, long unsigned int>, boost::_bi::list3<boost::_bi::value<TURAG::Feldbus::SerialFieldbus*>, boost::arg<1>, boost::arg<2> > >&; typename boost::asio::async_result<typename std::decay<WriteHandler>::type, void(boost::system::error_code, long unsigned int)>::return_type = void; typename std::enable_if<boost::asio::is_mutable_buffer_sequence<MutableBufferSequence>::value>::type = void]’:
/home/hermann/turag/programmierung/ros/src/turag_tina/tina/platform/ros/serialfieldbus.cpp:52:76: warning: zero as null pointer constant [-Wzero-as-null-pointer-constant]
   52 |         async_read(port_, boost::asio::buffer(receive_, receive_length_), f);
      |                                                                            ^
/home/hermann/turag/programmierung/ros/src/turag_tina/tina/platform/ros/serialfieldbus.cpp: In function ‘typename boost::asio::async_result<typename std::decay<WriteHandler>::type, void(boost::system::error_code, long unsigned int)>::return_type boost::asio::async_write(AsyncWriteStream&, const ConstBufferSequence&, WriteHandler&&, typename std::enable_if<boost::asio::is_const_buffer_sequence<ConstBufferSequence>::value>::type*) [with AsyncWriteStream = boost::asio::basic_serial_port<>; ConstBufferSequence = boost::asio::const_buffers_1; WriteHandler = boost::_bi::bind_t<void, boost::_mfi::mf2<void, TURAG::Feldbus::SerialFieldbus, const boost::system::error_code&, long unsigned int>, boost::_bi::list3<boost::_bi::value<TURAG::Feldbus::SerialFieldbus*>, boost::arg<1>, boost::arg<2> > >&; typename boost::asio::async_result<typename std::decay<WriteHandler>::type, void(boost::system::error_code, long unsigned int)>::return_type = void; typename std::enable_if<boost::asio::is_const_buffer_sequence<ConstBufferSequence>::value>::type = void]’:
/home/hermann/turag/programmierung/ros/src/turag_tina/tina/platform/ros/serialfieldbus.cpp:105:77: warning: zero as null pointer constant [-Wzero-as-null-pointer-constant]
  105 |         async_write(port_, boost::asio::buffer(transmit,transmit_length_), f);
      |                                                                             ^
```

This is because https://github.com/chriskohlhoff/asio/blob/master/asio/include/asio/read.hpp and https://github.com/chriskohlhoff/asio/blob/master/asio/include/asio/write.hpp set `0` as default value for various parameters with `std::enable_if`.

These methods should either use `enable_if<..., int>::type = 0` as type or set the default value to `nullptr` if it is available.

**Comment by stertingen** [comment]
Related:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=65324
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=52718

This is a GCC bug. Looking at how old this is, I wonder whether this is going to be fixed at all.

**Comment by chriskohlhoff** [comment]
Fixed in asio 1.18.1 / boost 1.75 by adding warning suppression to asio/detail/push_options.hpp.
